Comprehensive IPv6 framework with support for freedns.

dkerr64 opened the pull request:
Updates here are based on work I have done for the AstLinux project (AstLinux.org)

Adding comprehensive framework for IPv6 support.  IPv6 and IPv4 exist side-by-side
The same network interface may have both IPv4 and IPv6.  Therefore to support both
we add the following options.

 usev6 = [ ip, if, cmd, web ]
 ipv6 = [ set an IPv6 address (usev6=ip) ]
 webv6 = [ url of a web server that will return your IPv6 address (usev6=web)]
 webv6-skip = [ string to skip to before looking for IPv6 address]
 cmdv6 = [ system command to execute that will return an IPv6 address (usev6=cmd)]
 cmdv6-skip = [ string to skip to before looking for IPv6 address ]

There is no ifv6. It is assumed that if=interface will be the same for both IPv4
and IPv6

You can mix methods for IPv4 and IPv6 for the same interface.  For example
 use=web
 usev6=if
 if=eth0
 web=http://ipdetect.dnspark.com/
Would be a valid method for a system behind an IPv4 NAT connection, IPv6 would be
obtained from the eth0 interface (globaly routable address) and the public IPv4
from external web site.

Support for updating AAAA (IPv6) records is added for freedns.afraid.org.  The
existing IPv6 nsupdate method was updated to account for new design.

Optional support is added for cURL as some embedded systems do not have perl
NET6 and SSL packages installed.  Set with option
 curl=yes

Submitted code is ready for code review and comment.

dkerr64 committed:
Updates here are based on work I have done for the AstLinux project (AstLinux.org)

Adding comprehensive framework for IPv6 support.  IPv6 and IPv4 exist side-by-side
The same network interface may have both IPv4 and IPv6.  Therefore to support both
we add the following options.

 usev6 = [ ip, if, cmd, web ]
 ipv6 = [ set an IPv6 address (usev6=ip) ]
 webv6 = [ url of a web server that will return your IPv6 address (usev6=web)]
 webv6-skip = [ string to skip to before looking for IPv6 address]
 cmdv6 = [ system command to execute that will return an IPv6 address (usev6=cmd)]
 cmdv6-skip = [ string to skip to before looking for IPv6 address ]

There is no ifv6. It is assumed that if=interface will be the same for both IPv4
and IPv6

You can mix methods for IPv4 and IPv6 for the same interface.  For example
 use=web
 usev6=if
 if=eth0
 web=http://ipdetect.dnspark.com/
Would be a valid method for a system behind an IPv4 NAT connection, IPv6 would be
obtained from the eth0 interface (globaly routable address) and the public IPv4
from external web site.

Support for updating AAAA (IPv6) records is added for freedns.afraid.org.  The
existing IPv6 nsupdate method was updated to account for new design.

Optional support is added for cURL as some embedded systems do not have perl
NET6 and SSL packages installed.  Set with option
 curl=yes

Submitted code is ready for code review and comment.

abelbeck commented on ddclient:
```diff
@@ -2075,6 +2176,37 @@ sub geturl {
+            $reply = `/usr/bin/curl -si0 --user "${login}:${password}" --user-agent "${program}/${version}" \\
+              --connect-timeout $timeout --max-time $timeout $curlopt \\
+              --url "${protocol}://${server}/${url}" 2>/dev/null`;
+
```
Excellent idea to optionally use ```curl``` as it is ubiquitous, performs either IPv4/IPv6 and SSL all in one call.

But, may I suggest you use "HERE Documents" to minimize unwanted special-character processing:
````
	$reply = <<`USE_CURL`;
/usr/bin/curl -si0 --user '${login}:${password}' --user-agent '${program}/${version}' \\
	  --connect-timeout $timeout --max-time $timeout $curlopt \\
	  --url '${protocol}://${server}/${url}' 2>/dev/null
USE_CURL
```
Also use singe-quotes around the string arguments, better to allow ${password} special characters and does not effect the ${foo} perl replacement.

dkerr64 commented:
Lonnie, thanks for the comments.  I am not familiar with the HERE Documents method but based on your comments it does look like being a good idea.  

Thanks

dkerr64 commented:
Further testing has revealed that the Perl built-in NET6 support is not a reliable method.  While it works fine if you are on a system that is IPv6 connected, it is failing in some cases if you try and use it on a system that is only IPv4.  ie, running "ddclient -query" fails half way through testing webv6 sites if you are on a system that is not currently connected IPv6.  It simply crashes with no diag info on the syswrite command ..
`my $result = syswrite $sd, $rq;`

One argument would of course be that you simply shouldn't do that... use INET6 on a system not IPv6 enabled.  But a simpler approach might just be to enforce use of Curl when IPv6 is requested.

David

dkerr64 committed:
Force use of cURL when IPv6 protocol requested.  The Perl NET6 functions do not fail
gracefully or fall back on IPv4 when running on a system that is not connected to IPv6
network.  Result is that ddclient may crash if IPv6 requested on a system that doesn't
have IPv6.

dkerr64 committed:
use "HERE Document" method for calling cURL command.

dkerr64 committed:
Merge remote-tracking branch 'upstream/master'

dkerr64 committed:
Fixed bug that would cause IPv4 address to be updated even if it was unchanged.
Caused by testing for IPv6 address change even though no IPv6 update was requested.
Added capability to specify whether IPv4 or IPv6 should be updated or not with new
values:
use=no
usev6=no

dkerr64 committed:
Fixed bugs in logic testing when to update IPv6 address

dkerr64 committed:
Fix syntax errors testing for verbose option to print status messages. Woops.

andig commented:
Whats the status of IPv6 support? I have a small raspi with a global IPv6 address that I'd like to tie to my domain?

dkerr64 commented:
@andig  I am using the IPv6 support I added with freedns.afraid.org and it is working fine.  I use it on two systems... my gateway, which obtains the IPv6 address from the ethernet interface and I use it on an internal server (behind IPv4 NAT) which obtains IPv6 addresses for several internal servers and updates them to freedns (if you want the outside world to get to those servers then you also need to add appropriate firewall rules to the gateway).

However it works only with the freedns service.  I have not and don't plan to try and make it work with any other service.  I leave that to someone else.

@wimpunk has not commented on the changes I submitted and has not merged into the master.  I would very much like it merged but that is not up to me.  I would highly recommend a thorough code review -- there are a lot of code changes and as you can see from the commit history I'm not perfect... there have been bugs that I found and corrected.  So, I am keen for other eyes to look this over.

Thanks your interest and support.

dak.

dkerr64 commented:
by-the-way.... in the above post I stated that I have one server that obtains IPv6 addresses for other servers on my internal network and updates the address to freedns.   How do I do that?  With the following ugly cmdv6= in ddclient.conf...
```
usev6=cmd
cmdv6="{ host=qnap; ping $host -c 1 >/dev/null; ipv6=$(ip addr show dev ens3 | grep \"scope.global\" | sed -e's/^.*inet6 \\([^ ]*\\)\\/.*$/\\1/;t;d'); mac=$(arp -a $host | sed -e's/^.* at \\([^ ]*\\) .*$/\\1/;t;d'); ipv6calc -q --in prefix+mac --action prefixmac2ipv6 --out ipv6addr  $ipv6 $mac 2>/dev/null; }"
```
In this case the internal server has a hostname of "qnap"
If you want to try this on a command line then be sure to remove the backslash's before the quotes and backlashes.  In other words...

`{ host=qnap; ping $host -c 1 >/dev/null; ipv6=$(ip addr show dev ens3 | grep "scope.global" | sed -e's/^.*inet6 \([^ ]*\)\/.*$/\1/;t;d'); mac=$(arp -a $host | sed -e's/^.* at \([^ ]*\) .*$/\1/;t;d'); ipv6calc -q --in prefix+mac --action prefixmac2ipv6 --out ipv6addr  $ipv6 $mac 2>/dev/null; }`

Try that from your linux cmd line (substitute valid local host name for qnap) !

dak

wimpunk commented:
Sorry I missed this one... I'll try to take a review asap.
Did anyone tried the patch already?

dkerr64 committed:
Merge remote-tracking branch 'upstream/master'

wimpunk commented:
Seems like there are issues when trying to merge and I have some reviews which didn't got any respons.  Can you review them?

dkerr64 committed:
Merge remote-tracking branch 'upstream/master'

dkerr64 commented:
Working on re-syncing with your current master, don't merge until I complete and test please.  I'll let you know once that is done.
Please point me at the review/comments that I have not responded to.

dkerr64 committed:
Improve IPv6 parsing and some error messages

wimpunk commented:
About the comments: there was a `load_ssl_support` removed from line 1981 and the identification changed.  I argued that I'm aware of a bad indentation but there is no need to correct it on rows you didn't change the contents on.

dkerr64 committed:
fix whitespace diffs

dkerr64 commented:
I have fixed all the white space diffs.

dkerr64 commented on ddclient:
```diff
@@ -1991,74 +2091,67 @@ sub geturl {
     if ( $force_ssl || ($globals{'ssl'} and (caller(1))[3] ne 'main::get_ip') ) {
         $use_ssl      = 1;
         $default_port = 443;
-		load_ssl_support;
```
This `load_ssl_support` is deleted here and moved down to where it is needed.  IPv6 only works with cURL so this should be loaded only if required for IPv4 and only if cURL is not being used for HTTP.

dkerr64 commented on ddclient:
```diff
@@ -1991,74 +2091,67 @@ sub geturl {
+	       debug("skipped network connection");
+	       verbose("SENDING:", "%s", $request);
+        } elsif ($use_ssl) {
+		load_ssl_support;
```
`load_ssl_suppor` moved here... where required.

dkerr64 commented:
`load_ssl_support` was moved lower into the code.  See inline comments.

dkerr64 commented:
An additional comment on `load_ssl_support`.  We ran into several problems using the perl IO::Socket with IPv6 and SSL eventually I had to abandon using it and so if you use IPv6 in ddclient (this pull request) then you need to use cURL... which has to be enabled in the .conf file -- which enables it for both IPv6 and IPv4.  After further review we decided to abandon IO:Socket completely and for the astinux-project pulled all the code out of ddclient in our own fork (https://github.com/astlinux-project/ddclient-curl).   cURL is much more robust and portable across systems and has proven much more stable.  With this pull request I have had to weave in IPv6 support using cURL without breaking any of your existing IPv4 code that uses IO:Socket making the end result more complicated than it need be.

dkerr64 committed:
Pass IPv6 address to postscript application (if present)

erul92 commented on ddclient:
```diff
@@ -4149,48 +4415,87 @@ sub nic_nsupdate_update {
+			}
+
 		}
 	}
```
load_ssl_support

nirgal commented:
Hi!

This is great work. I really need ipv4+ipv6 support for http://njalladnspotetti.onion dyndns service: It requires setting up ipv4 and ipv6 at the same time; Setting only one will remove the other.

Did you try to create a push request upstream? (source forge)

dkerr64 commented:
@nirgal see here... https://github.com/astlinux-project/ddclient-curl
Unfortunately @abelbeck and I had to create our own fork for our project as @wimpunk did not act on our PR and we needed this and other changes.

nirgal commented:
I'm reviewing the PR, in case wimpunk wakes up. So far my only interrogation is regarding the non-default curl option "--insecure". I'm not sure I want that in my ddclient. Can you comment on this? I suppose this disables certificate checks and this is required for some routers? How is that different from situation before your commit?

dkerr64 commented:
@nirgal you are correct, this flag stops CURL from checking that the certificate is valid.  I don't recall exactly why I added this but I assume it is because I ran into cases where we are communicating with sites that have self signed certificates.  It would not be a bad idea to make this an option so that it is only turned on if the user runs into an issue and needs it.

Duffyx commented:
Hello, has this been merged into the repository? As the PR is still open. I would very much like decent IPv6 support on a dynamic DNS client. Thanks!

dkerr64 commented:
We created our own fork because ddclient owner never merged this.  See here...
https://github.com/astlinux-project/ddclient-curl

Thanks
DAK

dominikholler commented:
Until ddclient supports dual-stack, I use ddclient's postscript hook to update the IPv6 by curl manually, see https://gist.github.com/dominikholler/aec112a2ee05288588895120f41189d7
Maybe this workaround helps other people, too.

ffries commented:
It is sad to see that IPv6 is not ddclient priority. 
Is the project dead?

vfreex commented:
+1 for this feature.
Most Chinese ISPs are supporting IPv6 now but unfortunately we don't have enough ddns support yet.

SuperSandro2000 commented:
I solved the merge conflicts with this branch here https://github.com/SuperSandro2000/ddclient .

Please test it as it could be buggy.

Also @wimpunk @DaveSophoServices if you don't have enough time to maintain this project I would love to help you and get all this stuff reviewed and merged.

wimpunk commented:
@SuperSandro2000 
I just saw you want to help us supporting ddclient. Are you already subscribed to the mailing list at https://sourceforge.net/p/ddclient/mailman/? Let's discuss what you can do on that list.

dachaac commented:
FYI:
Just received a newsletter from no-ip.com and they are now finally providing IPv6 support for their service. They claim that they are pretty close to "industry standard". But I suppose you know that better.

Ref:
https://www.noip.com/integrate
https://www.noip.com/integrate/request
https://www.noip.com/integrate/add-noip

Doesn't look too complicated once base IPv6 support is in place. Is there some help needed to get this in or just waiting it to happen?

dkerr64 commented:
There are too many merge conflicts with the current version of ddclient.  I will bring the IPv6 support up-to-date with current master and then resubmit a new pull request.  You can follow along here if you are interested.... https://github.com/dkerr64/ddclient/tree/IPv6
Thanks.
